In [1]:
import pandas as pd
import numpy as np

pd.set_option('display.max_columns',None)
pd.set_option('display.max_rows',None)

pd.set_option('display.float_format', lambda x: '%.2f' % x)

In [3]:
#2010-2011 yılı içerisindeki veriler
df_2010_2011 = pd.read_excel('online_retail_II.xlsx',sheet_name="Year 2010-2011")

In [37]:
df = df_2010_2011.copy()

In [6]:
df.head()

Invoice StockCode                          Description  Quantity  \
0  536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1  536365     71053                  WHITE METAL LANTERN         6   
2  536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3  536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4  536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  Price  Customer ID         Country  
0 2010-12-01 08:26:00   2.55     17850.00  United Kingdom  
1 2010-12-01 08:26:00   3.39     17850.00  United Kingdom  
2 2010-12-01 08:26:00   2.75     17850.00  United Kingdom  
3 2010-12-01 08:26:00   3.39     17850.00  United Kingdom  
4 2010-12-01 08:26:00   3.39     17850.00  United Kingdom

In [7]:
df.isnull().sum()

Invoice             0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
Price               0
Customer ID    135080
Country             0
dtype: int64

In [8]:
df['Description'].nunique()

4223

In [9]:
df['Description'].value_counts().head()

WHITE HANGING HEART T-LIGHT HOLDER    2369
REGENCY CAKESTAND 3 TIER              2200
JUMBO BAG RED RETROSPOT               2159
PARTY BUNTING                         1727
LUNCH BAG RED RETROSPOT               1638
Name: Description, dtype: int64

In [12]:
#en çok sipariş edilen ürün hangisidir?
df.groupby('Description').agg({"Quantity":"sum"}).sort_values('Quantity', ascending = False).head()

Quantity
Description                                
WORLD WAR 2 GLIDERS ASSTD DESIGNS     53847
JUMBO BAG RED RETROSPOT               47363
ASSORTED COLOUR BIRD ORNAMENT         36381
POPCORN HOLDER                        36334
PACK OF 72 RETROSPOT CAKE CASES       36039

In [38]:
df["TotalPrice"] = df["Quantity"]*df["Price"]

In [14]:
df.head()

Invoice StockCode                          Description  Quantity  \
0  536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1  536365     71053                  WHITE METAL LANTERN         6   
2  536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3  536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4  536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  Price  Customer ID         Country  TotalPrice  
0 2010-12-01 08:26:00   2.55     17850.00  United Kingdom       15.30  
1 2010-12-01 08:26:00   3.39     17850.00  United Kingdom       20.34  
2 2010-12-01 08:26:00   2.75     17850.00  United Kingdom       22.00  
3 2010-12-01 08:26:00   3.39     17850.00  United Kingdom       20.34  
4 2010-12-01 08:26:00   3.39     17850.00  United Kingdom       20.34

In [20]:
df["Description"].value_counts().head()

Manual                      244
REGENCY CAKESTAND 3 TIER    181
POSTAGE                     126
JAM MAKING SET WITH JARS     87
Discount                     77
Name: Description, dtype: int64

In [21]:
df.head()

Invoice StockCode                       Description  Quantity  \
141  C536379         D                          Discount        -1   
154  C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235  C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236  C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237  C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   

            InvoiceDate  Price  Customer ID         Country  TotalPrice  
141 2010-12-01 09:41:00  27.50     14527.00  United Kingdom      -27.50  
154 2010-12-01 09:49:00   4.65     15311.00  United Kingdom       -4.65  
235 2010-12-01 10:24:00   1.65     17548.00  United Kingdom      -19.80  
236 2010-12-01 10:24:00   0.29     17548.00  United Kingdom       -6.96  
237 2010-12-01 10:24:00   0.29     17548.00  United Kingdom       -6.96

In [39]:
df.dropna(inplace = True)

In [40]:
df.isnull().sum()

Invoice        0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
Price          0
Customer ID    0
Country        0
TotalPrice     0
dtype: int64

In [41]:
df.shape

(406830, 9)

In [42]:
df.describe([0.01,0.05,0.10,0.25,0.50,0.75,0.90,0.95,0.99]).T

count     mean     std        min       1%       5%      10%  \
Quantity    406830.00    12.06  248.69  -80995.00    -2.00     1.00     1.00   
Price       406830.00     3.46   69.32       0.00     0.21     0.42     0.55   
Customer ID 406830.00 15287.68 1713.60   12346.00 12415.00 12626.00 12876.00   
TotalPrice  406830.00    20.40  427.59 -168469.60    -9.90     0.85     1.65   

                 25%      50%      75%      90%      95%      99%       max  
Quantity        2.00     5.00    12.00    24.00    36.00   120.00  80995.00  
Price           1.25     1.95     3.75     6.75     8.50    15.00  38970.00  
Customer ID 13953.00 15152.00 16791.00 17719.00 17905.00 18212.00  18287.00  
TotalPrice      4.20    11.10    19.50    35.40    67.20   199.68 168469.60

# Recency

In [43]:
df.head()

Invoice StockCode                          Description  Quantity  \
0  536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1  536365     71053                  WHITE METAL LANTERN         6   
2  536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3  536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4  536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  Price  Customer ID         Country  TotalPrice  
0 2010-12-01 08:26:00   2.55     17850.00  United Kingdom       15.30  
1 2010-12-01 08:26:00   3.39     17850.00  United Kingdom       20.34  
2 2010-12-01 08:26:00   2.75     17850.00  United Kingdom       22.00  
3 2010-12-01 08:26:00   3.39     17850.00  United Kingdom       20.34  
4 2010-12-01 08:26:00   3.39     17850.00  United Kingdom       20.34

In [44]:
import datetime as dt

In [45]:
df["InvoiceDate"].min()

Timestamp('2010-12-01 08:26:00')

In [46]:
df["InvoiceDate"].max()

Timestamp('2011-12-09 12:50:00')

In [47]:
today_date = dt.datetime(2011,12,9)

In [49]:
df["Customer ID"] = df["Customer ID"].astype(int)

In [50]:
df.head()

Invoice StockCode                          Description  Quantity  \
0  536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1  536365     71053                  WHITE METAL LANTERN         6   
2  536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3  536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4  536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  Price  Customer ID         Country  TotalPrice  
0 2010-12-01 08:26:00   2.55        17850  United Kingdom       15.30  
1 2010-12-01 08:26:00   3.39        17850  United Kingdom       20.34  
2 2010-12-01 08:26:00   2.75        17850  United Kingdom       22.00  
3 2010-12-01 08:26:00   3.39        17850  United Kingdom       20.34  
4 2010-12-01 08:26:00   3.39        17850  United Kingdom       20.34

In [51]:
df.groupby("Customer ID").agg({"InvoiceDate":"max"}).head()

InvoiceDate
Customer ID                    
12346       2011-01-18 10:17:00
12347       2011-12-07 15:52:00
12348       2011-09-25 13:13:00
12349       2011-11-21 09:51:00
12350       2011-02-02 16:01:00

In [52]:
today_date - df.groupby("Customer ID").agg({"InvoiceDate":"max"}).head()

InvoiceDate
Customer ID                  
12346       324 days 13:43:00
12347         1 days 08:08:00
12348        74 days 10:47:00
12349        17 days 14:09:00
12350       309 days 07:59:00

In [53]:
temp_df = (today_date - df.groupby("Customer ID").agg({"InvoiceDate":"max"}))

In [54]:
temp_df.head()

InvoiceDate
Customer ID                  
12346       324 days 13:43:00
12347         1 days 08:08:00
12348        74 days 10:47:00
12349        17 days 14:09:00
12350       309 days 07:59:00

In [55]:
temp_df.rename(columns={"InvoiceDate":"Recency"}, inplace = True)

In [56]:
temp_df.head()

Recency
Customer ID                  
12346       324 days 13:43:00
12347         1 days 08:08:00
12348        74 days 10:47:00
12349        17 days 14:09:00
12350       309 days 07:59:00

In [58]:
df.groupby("Customer ID").agg({"InvoiceDate":lambda x:(today_date-x.max()).days}).head()

InvoiceDate
Customer ID             
12346                324
12347                  1
12348                 74
12349                 17
12350                309

# Frequency

In [59]:
df.head()

Invoice StockCode                          Description  Quantity  \
0  536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1  536365     71053                  WHITE METAL LANTERN         6   
2  536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3  536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4  536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  Price  Customer ID         Country  TotalPrice  
0 2010-12-01 08:26:00   2.55        17850  United Kingdom       15.30  
1 2010-12-01 08:26:00   3.39        17850  United Kingdom       20.34  
2 2010-12-01 08:26:00   2.75        17850  United Kingdom       22.00  
3 2010-12-01 08:26:00   3.39        17850  United Kingdom       20.34  
4 2010-12-01 08:26:00   3.39        17850  United Kingdom       20.34

In [60]:
freq_df = df.groupby("Customer ID").agg({"InvoiceDate":"nunique"})

In [61]:
freq_df.head()

InvoiceDate
Customer ID             
12346                  2
12347                  7
12348                  4
12349                  1
12350                  1

In [62]:
freq_df.rename(columns={"InvoiceDate": "Frequency"}, inplace=True)

# Monetary

In [63]:
df.head()

Invoice StockCode                          Description  Quantity  \
0  536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1  536365     71053                  WHITE METAL LANTERN         6   
2  536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3  536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4  536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  Price  Customer ID         Country  TotalPrice  
0 2010-12-01 08:26:00   2.55        17850  United Kingdom       15.30  
1 2010-12-01 08:26:00   3.39        17850  United Kingdom       20.34  
2 2010-12-01 08:26:00   2.75        17850  United Kingdom       22.00  
3 2010-12-01 08:26:00   3.39        17850  United Kingdom       20.34  
4 2010-12-01 08:26:00   3.39        17850  United Kingdom       20.34

In [64]:
monetary_df = df.groupby("Customer ID").agg({"TotalPrice":"sum"})

In [65]:
monetary_df.head()

TotalPrice
Customer ID            
12346              0.00
12347           4310.00
12348           1797.24
12349           1757.55
12350            334.40

In [66]:
monetary_df.rename(columns={"TotalPrice":"Monetary"}, inplace=True)

In [67]:
monetary_df.head()

Monetary
Customer ID          
12346            0.00
12347         4310.00
12348         1797.24
12349         1757.55
12350          334.40

In [68]:
print(recency_df.shape,freq_df.shape,monetary_df.shape)

(4372,) (4372, 1) (4372, 1)


In [69]:
rfm = pd.concat([recency_df, freq_df, monetary_df],  axis=1)

In [70]:
rfm.head()

Recency  Frequency  Monetary
Customer ID                              
12346            324          2      0.00
12347              1          7   4310.00
12348             74          4   1797.24
12349             17          1   1757.55
12350            309          1    334.40

In [71]:
rfm["RecencyScore"] = pd.qcut(rfm["Recency"], 5, labels = [5, 4, 3, 2, 1])
rfm["FrequencyScore"] = pd.qcut(rfm["Frequency"].rank(method="first"), 5, labels = [1, 2, 3, 4, 5])
rfm["MonetaryScore"] = pd.qcut(rfm["Monetary"].rank(method="first"), 5, labels = [1, 2, 3, 4, 5])

In [72]:
rfm.head()

Recency  Frequency  Monetary RecencyScore FrequencyScore  \
Customer ID                                                             
12346            324          2      0.00            1              2   
12347              1          7   4310.00            5              4   
12348             74          4   1797.24            2              3   
12349             17          1   1757.55            4              1   
12350            309          1    334.40            1              1   

            MonetaryScore  
Customer ID                
12346                   1  
12347                   5  
12348                   4  
12349                   4  
12350                   2

In [73]:
(rfm["RecencyScore"].astype(str)+
rfm["FrequencyScore"].astype(str)+
rfm["MonetaryScore"].astype(str)).head()

Customer ID
12346    121
12347    545
12348    234
12349    414
12350    112
dtype: object

In [74]:
# rfm skorları kategorik değere dönüştürülüp df'e eklendi
rfm["RFM_SCORE"] = (rfm['RecencyScore'].astype(str) + 
                    rfm['FrequencyScore'].astype(str) + 
                    rfm['MonetaryScore'].astype(str))

In [75]:
rfm.head()

Recency  Frequency  Monetary RecencyScore FrequencyScore  \
Customer ID                                                             
12346            324          2      0.00            1              2   
12347              1          7   4310.00            5              4   
12348             74          4   1797.24            2              3   
12349             17          1   1757.55            4              1   
12350            309          1    334.40            1              1   

            MonetaryScore RFM_SCORE  
Customer ID                          
12346                   1       121  
12347                   5       545  
12348                   4       234  
12349                   4       414  
12350                   2       112

In [76]:
# Regular Expressions (Düzenli İfadeler) kullanılarak RFM haritası çıkarıldı
seg_map = {
    r'[1-2][1-2]': 'Hibernating',
    r'[1-2][3-4]': 'At Risk',
    r'[1-2]5': 'Can\'t Loose',
    r'3[1-2]': 'About to Sleep',
    r'33': 'Need Attention',
    r'[3-4][4-5]': 'Loyal Customers',
    r'41': 'Promising',
    r'51': 'New Customers',
    r'[4-5][2-3]': 'Potential Loyalists',
    r'5[4-5]': 'Champions'
}

In [77]:
rfm['Segment'] = rfm['RecencyScore'].astype(str) + rfm['FrequencyScore'].astype(str)

In [78]:
rfm.head()

Recency  Frequency  Monetary RecencyScore FrequencyScore  \
Customer ID                                                             
12346            324          2      0.00            1              2   
12347              1          7   4310.00            5              4   
12348             74          4   1797.24            2              3   
12349             17          1   1757.55            4              1   
12350            309          1    334.40            1              1   

            MonetaryScore RFM_SCORE Segment  
Customer ID                                  
12346                   1       121      12  
12347                   5       545      54  
12348                   4       234      23  
12349                   4       414      41  
12350                   2       112      11

In [79]:
rfm['Segment'] = rfm['Segment'].replace(seg_map, regex=True)

In [80]:
rfm.head()

Recency  Frequency  Monetary RecencyScore FrequencyScore  \
Customer ID                                                             
12346            324          2      0.00            1              2   
12347              1          7   4310.00            5              4   
12348             74          4   1797.24            2              3   
12349             17          1   1757.55            4              1   
12350            309          1    334.40            1              1   

            MonetaryScore RFM_SCORE      Segment  
Customer ID                                       
12346                   1       121  Hibernating  
12347                   5       545    Champions  
12348                   4       234      At Risk  
12349                   4       414    Promising  
12350                   2       112  Hibernating

In [81]:
rfm[["Segment","Recency","Frequency", "Monetary"]].groupby("Segment").agg(["mean","median","count"])

Recency              Frequency              Monetary  \
                       mean median count      mean median count     mean   
Segment                                                                    
About to Sleep        51.10     51   370      1.29      1   370   435.29   
At Risk              156.54    140   606      3.31      3   606   948.96   
Can't Loose          133.85    106    66      9.80      8    66  2431.87   
Champions              4.08      3   665     14.45      9   665  6510.57   
Hibernating          216.19    217  1052      1.20      1  1052   376.14   
Loyal Customers       31.36     27   795      7.94      6   795  2706.46   
Need Attention        47.53     45   187      2.63      3   187   774.94   
New Customers          5.24      6    42      1.00      1    42   377.23   
Potential Loyalists   14.61     15   492      2.21      2   492   683.79   
Promising             21.10     21    97      1.00      1    97   303.06   

                                   
                     median count  
Segment                            
About to Sleep       327.20   370  
At Risk              632.78   606  
Can't Loose         1891.30    66  
Champions           2531.58   665  
Hibernating          266.69  1052  
Loyal Customers     1708.86   795  
Need Attention       598.83   187  
New Customers        246.81    42  
Potential Loyalists  486.30   492  
Promising            228.91    97

### Champions
- 665 müşteri
- Recency si en düşük olan yani en son satın alma sürelerinin ortalaması en düşük olan müşteri grubudur.
- Alışveriş sıklığı en fazla olan yani frequency si en yüksek olan müşteri grubudur.
- En fazla geliri sağlayan, monetary si en yüksek olan müşteri grubudur.
- Yeni ürün tanıtımında bu grup odak olabilir.
- Ödemeye isteklidirler, bu yüzden satışları artırmak için indirimli fiyatları bu gruba sunmayabiliriz.
- Ödeme konusunda çekinmiyorlar, kdv değeri yüksek ürünleri önerebiliriz. Bir ürünün en iyisini almak isteyeceklerdir.

### Can't Loose
- 66 müşteri
- Alışveriş sıklığı yüksek (Frequency)
- Sağladı gelir yüksek (Monetary) 
- Çok uzun süre önce alışveriş yapmış.
- Eskiden sık sık alışveriş yaparlardı geri döndürmek önemli.
- İndirim teklifleri geri dönmelerinde etkili olabilir.
- Önerilen ürünleri onlara göre kişileştirmek işe yarayabilir.

### Loyal Customers
- 795 müşteri
- Ortalama recency değerleri 27 dir.
- Alışveriş yapma sıklığı frequency 6.
- Sağladığı gelir yüksek.
- Ürünlerle ilgili geri dönüş alınabilir, memnuniyelerine göre kaybedilip kaybedilmeme durumu gözlenir ve olumsuz dönüşler için çözüm bulunabilir.
- Bu müşterilerin değerli olduğunu onlara hissettirmek gerekir. Hediyeler gönderilebilir. 

In [82]:
rfm[rfm["Segment"] == "Loyal Customers"].index # Loyal Customers müşteri numaralarını listele

Int64Index([12352, 12370, 12380, 12388, 12395, 12407, 12408, 12413, 12415,
            12421,
            ...
            18172, 18177, 18188, 18204, 18211, 18221, 18226, 18236, 18257,
            18263],
           dtype='int64', name='Customer ID', length=795)

In [84]:
new_df = pd.DataFrame()

In [85]:
new_df["NewCustomerID"] = rfm[rfm["Segment"] == "Loyal Customers"].index

In [87]:
new_df.head()

NewCustomerID
0          12352
1          12370
2          12380
3          12388
4          12395

In [91]:
new_df.to_csv("loyal_customers.csv") # df'i kaydet

In [92]:
new_df.to_csv('loyal_customers.csv', index=False)